### PR TITLE
Persist simple diskann index

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: protos/index.proto [hunk 1]
```diff
@@ -108,6 +108,9 @@ message DiskAnn {
 
   // L parameter
   uint32 L = 5;
+
+  /// Entry points to the graph
```
Comment (changhiskhan) [inline protos/index.proto:112]: these are the vertex id's?

File: python/python/lance/dataset.py [hunk 2]
```diff
@@ -402,14 +404,15 @@ def create_index(
         ]:
             raise ValueError(f"Metric {metric} not supported.")
         index_type = index_type.upper()
-        if index_type != "IVF_PQ":
+        if index_type not in ["IVF_PQ", "DISKANN"]:
             raise NotImplementedError(
                 f"Only IVF_PQ index_type supported. Got {index_type}"
```
Comment (changhiskhan) [inline python/python/lance/dataset.py:409]: change the message to include DISKANN?

File: python/src/dataset.rs [hunk 3]
```diff
@@ -275,39 +278,72 @@ impl Dataset {
...
+        let params = match index_type.to_uppercase().as_str() {
+            "IVF_PQ" => {
+                let mut ivf_params = IvfBuildParams::default();
+                let mut pq_params = PQBuildParams::default();
+                if let Some(kwargs) = kwargs {
+                    if let Some(n) = kwargs.get_item("num_partitions") {
+                        ivf_params.num_partitions = PyAny::downcast::<PyInt>(n)?.extract()?
+                    };
+
+                    if let Some(n) = kwargs.get_item("num_bits") {
```
Comment (changhiskhan) [inline python/src/dataset.rs:305]: i forgot do we actually support configuring this now?

File: python/src/dataset.rs [hunk 4]
```diff
@@ -275,39 +278,72 @@ impl Dataset {
...
+
+                    if let Some(o) = kwargs.get_item("max_opq_iterations") {
+                        pq_params.max_opq_iters = PyAny::downcast::<PyInt>(o)?.extract()?
+                    };
+                }
+                VectorIndexParams::with_ivf_pq_params(m_type, ivf_params, pq_params)
+            }
+            "DISKANN" => {
+                let mut params = DiskANNParams::default();
+                if let Some(kwargs) = kwargs {
```
Comment (changhiskhan) [inline python/src/dataset.rs:325]: what're the default values for these? prolly worth tuning it and set reasonable defaults for like high embedding dimensions and like up to 1M vectors ?

**Review by changhiskhan — APPROVED** [review]

File: python/src/dataset.rs (hunk 3 above)
Comment (eddyxu) [reply python/src/dataset.rs:305]: We dont support "throw error for unsupported params". This is just do the handing silently. I am fine to remove it if you feel strongly.

File: protos/index.proto (hunk 1 above)
Comment (eddyxu) [reply protos/index.proto:112]: These are the IDs in the graph. Basically the "index" in the graph file. Not the "row id" in original dataset

File: python/python/lance/dataset.py (hunk 2 above)
Comment (eddyxu) [reply python/python/lance/dataset.py:409]: Sure, will do.

File: python/src/dataset.rs (hunk 4 above)
Comment (eddyxu) [reply python/src/dataset.rs:325]: The default values were from the diskann paper. Will adjust them if necessary when we benchmark diskann

**Pull request merged**